### DISCUSSION TITLE
Cannot get sendgrid plugin to send emails

### DISCUSSION THREAD

**Discussion by star-ot** [medusajs/medusa#4291] [Q&A]
Hi all, my team and I have followed all documentation regarding setting up sendgrid, but are unable to get emails to send for any notifications that are subscribed.

We are using the local event bus for testing and events seem to be processing fine, but the actual dispatch of emails via sendgrid is not working. When the plugin is initalized, it appears to be taking the options from our .env and loading three times then the options go empty. ( I tested by adding some prints to console when the sendgrid.js is initialized with the _mail["default"].setApiKey ).

In order to make sure the @sendgrid node module was actually working, I sent my own test message in the same initialization function to myself and my co-worker and it worked fine when the plugin initialized. (See commented code below)

Can anyone help us figure this out? Is this an actual issue or are we just doing it wrong? We are using the fake redis and pretty much a clean install of the backend with postgres.

Again, it appears the sendgridservice is loaded 3 times as shown by my output here, but does not actually connect with the right api key or options?

TIA!

medusa-plugin-sendgrid: 1.3.9
Node.js v18.16.0
Postgres
Linux
Opera GX / Google Chrome / Firefox

## medusa develop output:
```powershell
⠋ Initializing modules
✔ Modules initialized – 6ms
✔ Database initialized – 96ms
✔ Repositories initialized – 34ms
✔ Services initialized – 69ms
✔ Express intialized – 9ms
⠙ Initializing plugins
SendGridService loading
SendGridService loaded
SendGridService options {
  api_key: 'SG.###',
  from: 'support@test.com',
  order_placed_template: 'd-705a1e2870c041d8afac7f31682cd544'
✔ Plugins intialized – 757ms
✔ Subscribers initialized – 8ms
✔ API initialized – 43ms
⠋ Initializing defaults
SendGridService loading
SendGridService loaded
SendGridService options {
  api_key: 'SG.###',
  from: 'support@test.com',
  order_placed_template: 'd-705a1e2870c041d8afac7f31682cd544'
}
SendGridService loading
SendGridService loaded
SendGridService options {
  api_key: 'SG.###'
}
✔ Defaults initialized – 62ms
✔ Indexing event emitted – 4ms
✔ Server is ready on port: 9000 – 13ms
```

## event bus output:
```powershell
info:    Processing order.placed which has 2 subscribers
```

## modified sendgrid.js:
```js
var SendGridService = /*#__PURE__*/function (_NotificationService) {
  _inherits(SendGridService, _NotificationService);

  var _super = _createSuper(SendGridService);

  /**
   * @param {Object} options - options defined in `medusa-config.js`
   *    e.g.
   *    {
   *      api_key: SendGrid api key
   *      from: Medusa <hello@medusa.example>,
   *      order_placed_template: 01234,
   *      order_updated_template: 56789,
   *      order_canceled_template: 4242,
   *      user_password_reset_template: 0000,
   *      customer_password_reset_template: 1111,
   *    }
   */
  function SendGridService(_ref, options) {
    // print something to make sure the service is loaded
    console.log("SendGridService loading"); // set the options
    const sgMail = require('@sendgrid/mail'); // testing.

    sgMail.setApiKey(process.env.SENDGRID_API_KEY);
    /*
    const msg = { // when I send my email using this method, it works fine.
      to: 'myself@test.com', 
      from: 'support@test.com', 
      subject: 'hi alex from medusajs server', 
      text: 'i am sending a message from medusajs server', 
      html: '<strong>work work</strong>', 
    }; 
    sgMail 
      .send(msg)  // testing.
      .then(() => console.log('send mail success')) 
      .catch(console.log);
      */
    var _this;

    var storeService = _ref.storeService,
        orderService = _ref.orderService,
        returnService = _ref.returnService,
        swapService = _ref.swapService,
        cartService = _ref.cartService,
        lineItemService = _ref.lineItemService,
        claimService = _ref.claimService,
        fulfillmentService = _ref.fulfillmentService,
        fulfillmentProviderService = _ref.fulfillmentProviderService,
        totalsService = _ref.totalsService,
        productVariantService = _ref.productVariantService,
        giftCardService = _ref.giftCardService;

    _classCallCheck(this, SendGridService);

    _this = _super.call(this);
    _this.options_ = options;
    _this.fulfillmentProviderService_ = fulfillmentProviderService;
    _this.storeService_ = storeService;
    _this.lineItemService_ = lineItemService;
    _this.orderService_ = orderService;
    _this.cartService_ = cartService;
    _this.claimService_ = claimService;
    _this.returnService_ = returnService;
    _this.swapService_ = swapService;
    _this.fulfillmentService_ = fulfillmentService;
    _this.totalsService_ = totalsService;
    _this.productVariantService_ = productVariantService;
    _this.giftCardService_ = giftCardService;
    options.api_key = process.env.SENDGRID_API_KEY;
    _mail["default"].setApiKey(process.env.SENDGRID_API_KEY); // i explicitly set the key here otherwise I got an error about the key not starting with "SG." when it clearly does.
    console.log("SendGridService loaded");
    console.log("SendGridService options", options);

    return _this;
  }
```

**Comment by pmltechpile** [comment]
Hi, did you ever figure out what was wrong? I seem to be experiencing the same issue.

**Reply by rkmcdonnell** [reply]
+1 - doesn't seem like anything is happening, using either the local or redis events module.  If anyone has suggestions, that would be much appreciated!

**Reply by cafrontadoe** [reply]
Hi, you have to create a sender verification in sendgrid>sender authentication>single sender verification.
then you can follow the oficial documentation on medusa to connect with sendgrid.
https://docs.medusajs.com/plugins/notifications/sendgrid#template-reference

**Reply by aldo-roman** [reply]
@cafrontadoe This is still happening to me with the single sender verified

**Comment by aldo-roman** [comment]
@star-ot @pmltechpile I found the root cause. Medusa 1.18 broke compatibility with the Sendgrid plugin, due to the different subscriber/loader architecture.

I've filed a bug report in #5969

**Comment by mohdziyadc** [comment]
I think this is because of the single sender verification email not being verified. I tried my personal mail id as a single sender and verified it. Then when the order was placed, a confirmation mail was sent from my personal mail to the customer's email. Using personal emails are not a good practice, but it's good enough for testing purposes. 

In your case, most probably the from address `support@test.com` wouldn't be verified. Apart from that I think the plugin is working correctly.

**Comment by NicolasGorga** [comment]
Hello, anyone solved this issue?

**Comment by lzn0301** [comment]
You can check if the event bus module is installed and configured. I did this and it worked.

https://docs.medusajs.com/development/events/modules/redis
